lib-src/libsndfile/man/sndfile-convert.1
.Dd November 2, 2014
.Dt SNDFILE-CONVERT 1
.Os
.Sh NAME
.Nm sndfile-convert
.Nd convert sound files from one format to another
.Sh SYNOPSIS
.Nm sndfile-convert
.Op Fl override-sample-rate Ns = Ns Ar rate
.Op Fl endian Ns = Ns Cm little | big | cpu
.Op Fl normalize
.Op Ar encoding
.Ar input
.Ar output
.Sh DESCRIPTION
.Nm
converts sound files from one audio format to another.
The output file is overwritten it it already exists.
.Ss Formats
The format of the output file is determined by the filename extension.
The following file formats are currently recognized:
.Pp
.Bl -tag -compact -width ircam
.It wav
WAV (Microsoft)
.It aif
AIFF (Apple/SGI)
.It au
AU (Sun/NeXT)
.It snd
AU (Sun/NeXT)
.It raw
RAW (header-less)
.It gsm
RAW (header-less)
.It vox
RAW (header-less)
.It paf
PAF (Ensoniq PARIS, big-endian)
.It fap
PAF (Ensoniq PARIS, little-endian)
.It svx
IFF (Amiga IFF/SVX8/SV16)
.It nist
SPHERE (NIST SPeech HEader Resources)
.It sph
SPHERE (NIST SPeech HEader Resources)
.It voc
VOC (Creative Labs)
.It ircam
SF (Berkeley/IRCAM/CARL)
.It sf
SF (Berkeley/IRCAM/CARL)
.It w64
W64 (SoundFoundry WAVE 64)
.It mat
MAT4 (GNU Octave 2.0 / Matlab 4.2)
.It mat4
MAT4 (GNU Octave 2.0 / Matlab 4.2)
.It mat5
MAT5 (GNU Octave 2.1 / Matlab 5.0)
.It pvf
PVF (Portable Voice Format)
.It xi
XI (FastTracker 2)
.It htk
HTK (HMM Tool Kit)
.It sds
SDS (Midi Sample Dump Standard)
.It avr
AVR (Audio Visual Research)
.It wavex
WAVEX (MS WAVE with WAVEFORMATEX)
.It sd2
SD2 (Sound Designer II)
.It flac
FLAC (FLAC Lossless Audio Codec)
.It caf
CAF (Apple Core Audio File)
.It wve
WVE (Psion Series 3)
.It prc
WVE (Psion Series 3)
.It ogg
OGG (OGG Container format)
.It oga
OGG (OGG Container format)
.It mpc
MPC (Akai MPC 2k)
.It rf64
RF64 (RIFF 64)
.El
.Ss Options
The following options are recoginzed:
.Pp
.Bl -tag -compact -width "override-sample-rate=XXXXX"
.It Fl override-sample-rate Ns = Ns Ar rate
Make the input use sample rate of
.Ar rate
Hz.
.It Fl endian Ns = Ns Cm little
Make the output file use little endian data.
.It Fl endian Ns = Ns Cm big
Make the output file use big endian data.
.It Fl endian Ns = Ns Cm cpu
Make the output file use CPU endianness.
.It Fl normalize
Normalize the audio data in the output file.
.El
.Ss Encodings
The optional
.Ar encoding
parameter allows setting of the data encoding for the output file.
The following encodings are currently supported:
.Pp
.Bl -tag -compact -width ima-adpcmXX
.It Fl pcms8
signed 8 bit pcm
.It Fl pcmu8
unsigned 8 bit pcm
.It Fl pcm16
16 bit pcm
.It Fl pcm24
24 bit pcm
.It Fl pcm32
32 bit pcm
.It Fl float32
32 bit floating point
.It Fl ulaw
ULAW
.It Fl alaw
ALAW
.It Fl ima-adpcm
IMA ADPCM (WAV only)
.It Fl ms-adpcm
MS ADPCM (WAV only)
.It Fl gsm610
GSM6.10 (WAV only)
.It Fl dwvw12
12 bit DWVW (AIFF only)
.It Fl dwvw16
16 bit DWVW (AIFF only)
.It Fl dwvw24
24 bit DWVW (AIFF only)
.It Fl vorbis
Vorbis (OGG only)
.El
.Pp
If no encoding is specified for the output file,
.Nm
will try to use the encoding of the input file.
This will not always work as most container formats
(e.g. WAV, AIFF etc) only support a small subset of encodings
(e.g. 16 bit PCM, a-law, Vorbis etc).
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Lk http://www.mega-nerd.com/libsndfile/
.Sh AUTHORS
.An Erik de Castro Lopo Aq Mt erikd@mega-nerd.com .


CHANGELOG.txt
List of significant changes and bug fixes going back to version 1.1.0


Changes in version 2.4.1:

Critical Bug Fix:
 * Audacity 2.4.0 was released with a critical bug, (Bug 2436),
   which is fixed in 2.4.1.  The bug caused loss of data when using two projects at the same time.

Improvements

 * Audacity now runs on Mac Catalina.
 * A new Time Toolbar showing current record/playback time has been
   split from the Selection Toolbar.  It can be resized large.
 * Presets for more effects can be imported and exported.
 * A new Multi-View mode for tracks.  In this mode both 
   spectrogram and waveform views are shown at the same time.
 * Opus available as a new audio export format on Windows and Linux.
 * Easier to convert labels between point labels and range labels.
 * New Loudness Normalization effect.
 * New RMS measurement analyzer.
 * New Noise Gate effect.
 * New Spectral Delete effect.

See also: https://wiki.audacityteam.org/wiki/New_features_in_Audacity_2.4.1


Bug Fixes

 Over 100 bugs in 2.3.3 fixed, including:

 * 2237 - Equalization effects have no Import or Export for curves
 * 2261 - Graphic EQ sliders have no frequency labels
 * 2265 - Track heights are reset to default on project re-opening
 * 2303 - Playback does not start at the saved cursor position on reopening a saved project.
 * 2314 - Zoom in fail when playing


Changes in version 2.3.3:

Improvements

 * Equalization effect now split into two effects, Filter Curve and Graphic EQ.
   * Presets (using manage button) now active/working.
   * Can now have two points at same frequency for steep steps.
 * 'What you hear is what you get' for exports.  Formerly the solo button 
   preferences could lead to differences in which tracks were exported.
 * Leading silence (blank space) not skipped over in exports.
 * Quality setting on AAC/M4A exports.
 * Some confusing functionality removed (better achieved in other ways)
   * Removed Nyquist Workbench (use built-in nyquist features)
   * Removed Vocal Remover (use Vocal Reduction)
   * Removed On-Demand aliased files (copy files instead)
   * Removed 'Normalize on Load' (Normalize as needed on export, instead)


Bug Fixes

 Over 150 bugs in 2.3.2 fixed, including:

 * 2112 - Crash when opening Equalization settings in a Macro
 * 2127 - Crash deleting a track while in Record/Pause state
 * 2176 - Crash when attempting to time shift multiple tracks
 * 2141 - Tabbing between labels stopped working
 * 2200 - Refusal to export some large (4GB) files on size grounds



Changes in version 2.3.2: 

Improvements

 * Audacity now includes the LAME mp3 encoder. Previously, due to now expired
   patents, you had to download it separately.
 * There is a new 'Select' button in the track panel to select the whole 
   track.
 * Audacity mod-script-pipe for driving Audacity from Python now comes with
   Audacity and can be enabled via preferences.
 * Options and preferences:
   * Type to Create a Label is now off by default.
 * A plug-in installer for Nyquist now provides a file browser for 
   selecting '.NY' files to install.

Bug Fixes

 Over 20 bugs in 2.3.1 fixed, including:

 * Audacity 2.3.1 would crash if append-recording to a collapsed track.
 * Many accessibility bugs on macOS fixed



Changes in version 2.3.1: 

Improvements

    * Audacity now supports 64 bit Mac.
    * Options and preferences:
        * Microfades are now an opt-in feature. They were always on in 2.3.0.
        * The advanced vertical zooming option is now available in the 
          View->Zoom menu.
    * Regular interval labels now support range labels.

Bug Fixes

 Over 20 bugs in 2.3.0 fixed, including:

    * A crashing bug on Linux that made Audacity 2.3.0. unsuitable for Linux.
    * Crash on Mac when right clicking on label.
    * Problems with Change Pitch, including not remembering parameter values.



Changes in version 2.3.0:

Improvements

    * New feature - "Punch and Roll Recording".
    * Pinned-play-head can now be repositioned by dragging
    * Play-at-Speed now can be adjusted whilst playing.
    * Toolbars controlling volume and speed can now be resized for greater 
      precision
    * Macros (formerly 'Chains') substantially extended
        * New Macro palette
        * Macros can be bound to keyboard keys
    * New commands
        * New 'Tools' menu
        * New 'Scriptables' commands 
    * Nyquist gains AUD-DO command
    * Nyquist effects are now translatable and translated
    * More dialogs have help buttons now
    * Increased legibility of trackname display
    * Half-wave option for collapsed tracks
    * Sliding Stretch
    * Dialog (option) for entering labels

Bug Fixes

    Over 90 bugs in 2.2.2 fixed, including:

    * Windows: Queen Mary 1.7.1 Vamp plug-ins crash Audacity on opening 
      Plug-in Manager - bug #1244
    * Inconsistent behavior when recording with a selection defined - bug #1839
    * Ctrl + M does not open label for editing - bug #1852
    * Vertical Zoom in by clicking in the vertical Scale, or context menu, in 
      waveform views is inconsistent - bug #1810
    * Deletion of all tracks cannot be undone - bug #1845
    * Timer Record with a selection present can get a truncated recording with 
      data loss - bug #1851
    * Duplicate shortcuts can be created - bug #1786
    * Labels longer than 260 characters are now supported - bug #1905
    * Using Spectrogram Settings in TCP or using Preferences causes Audacity 
      to reset Project Rate to default rate in Quality Preferences - bug #1977
    * Cut Preview should play all selected/sync-locked tracks, respecting 
      Mute/Solo during preview - bug #231



Changes in version 2.2.2:

Improvements

    * Dropout detection
      Dropout detection is controlled from a setting in Recording Preferences 
      called "Detect dropouts".
      When this setting is "on" (default setting) Audacity will detect dropouts 
      (brief gaps in the recording) and will insert zeroes into the recording 

lib-src/libsndfile/programs/sndfile-convert.c
/*
** Copyright (C) 1999-2019 Erik de Castro Lopo <erikd@mega-nerd.com>
**
** All rights reserved.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are
** met:
**
**     * Redistributions of source code must retain the above copyright
**       notice, this list of conditions and the following disclaimer.
**     * Redistributions in binary form must reproduce the above copyright
**       notice, this list of conditions and the following disclaimer in
**       the documentation and/or other materials provided with the
**       distribution.
**     * Neither the author nor the names of any contributors may be used
**       to endorse or promote products derived from this software without
**       specific prior written permission.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
** TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
** PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
** CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
** EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
** PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
** OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
** OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
** ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <sndfile.h>

#include "common.h"


typedef	struct
{	char	*infilename, *outfilename ;
	SF_INFO	infileinfo, outfileinfo ;
} OptionData ;

static void copy_metadata (SNDFILE *outfile, SNDFILE *infile, int channels) ;

static void
usage_exit (const char *progname)
{
	printf ("\nUsage : %s [options] [encoding] <input file> <output file>\n", progname) ;
	puts ("\n"
		"    where [option] may be:\n\n"
		"        -override-sample-rate=X  : force sample rate of input to X\n"
		"        -endian=little           : force output file to little endian data\n"
		"        -endian=big              : force output file to big endian data\n"
		"        -endian=cpu              : force output file same endian-ness as the CPU\n"
		"        -normalize               : normalize the data in the output file\n"
		) ;

	puts (
		"    where [encoding] may be one of the following:\n\n"
		"        -pcms8     : signed 8 bit pcm\n"
		"        -pcmu8     : unsigned 8 bit pcm\n"
		"        -pcm16     : 16 bit pcm\n"
		"        -pcm24     : 24 bit pcm\n"
		"        -pcm32     : 32 bit pcm\n"
		"        -float32   : 32 bit floating point\n"
		"        -float64   : 64 bit floating point\n"
		"        -ulaw      : ULAW\n"
		"        -alaw      : ALAW\n"
		"        -alac16    : 16 bit ALAC (CAF only)\n"
		"        -alac20    : 20 bit ALAC (CAF only)\n"
		"        -alac24    : 24 bit ALAC (CAF only)\n"
		"        -alac32    : 32 bit ALAC (CAF only)\n"
		"        -ima-adpcm : IMA ADPCM (WAV only)\n"
		"        -ms-adpcm  : MS ADPCM (WAV only)\n"
		"        -gsm610    : GSM6.10 (WAV only)\n"
		"        -dwvw12    : 12 bit DWVW (AIFF only)\n"
		"        -dwvw16    : 16 bit DWVW (AIFF only)\n"
		"        -dwvw24    : 24 bit DWVW (AIFF only)\n"
		"        -vorbis    : Vorbis (OGG only)\n"
		"        -opus      : Opus (OGG only)\n"
		) ;

	puts (
		"    If no encoding is specified, the program will try to use the encoding\n"
		"    of the input file in the output file. This will not always work as\n"
		"    most container formats (eg WAV, AIFF etc) only support a small subset\n"
		"    of codec formats (eg 16 bit PCM, a-law, Vorbis etc).\n"
		) ;

	puts (
		"    The format of the output file is determined by the file extension of the\n"
		"    output file name. The following extensions are currently understood:\n"
		) ;

	sfe_dump_format_map () ;

	puts ("") ;
	exit (1) ;
} /* usage_exit */

static void
report_format_error_exit (const char * argv0, SF_INFO * sfinfo)
{	int old_format = sfinfo->format ;
	int endian = sfinfo->format & SF_FORMAT_ENDMASK ;
	int channels = sfinfo->channels ;

	sfinfo->format = old_format & (SF_FORMAT_TYPEMASK | SF_FORMAT_SUBMASK) ;

	if (endian && sf_format_check (sfinfo))
	{	printf ("Error : output file format does not support %s endian-ness.\n", sfe_endian_name (endian)) ;
		exit (1) ;
		} ;

	sfinfo->channels = 1 ;
	if (sf_format_check (sfinfo))
	{	printf ("Error : output file format does not support %d channels.\n", channels) ;
		exit (1) ;
		} ;

	printf ("\n"
			"Error : output file format is invalid.\n"
			"The '%s' container does not support '%s' codec data.\n"
			"Run '%s --help' for clues.\n\n",
			sfe_container_name (sfinfo->format), sfe_codec_name (sfinfo->format), program_name (argv0)) ;
	exit (1) ;
} /* report_format_error_exit */

int
main (int argc, char * argv [])
{	const char	*progname, *infilename, *outfilename ;
	SNDFILE		*infile = NULL, *outfile = NULL ;
	SF_INFO		sfinfo ;
	int			k, outfilemajor, outfileminor = 0, infileminor ;
	int			override_sample_rate = 0 ; /* assume no sample rate override. */
	int			endian = SF_ENDIAN_FILE, normalize = SF_FALSE ;

	progname = program_name (argv [0]) ;

	if (argc < 3 || argc > 5)
		usage_exit (progname) ;

	infilename = argv [argc-2] ;
	outfilename = argv [argc-1] ;

	if (strcmp (infilename, outfilename) == 0)
	{	printf ("Error : Input and output filenames are the same.\n\n") ;
		usage_exit (progname) ;
		} ;

	if (strlen (infilename) > 1 && infilename [0] == '-')
	{	printf ("Error : Input filename (%s) looks like an option.\n\n", infilename) ;
		usage_exit (progname) ;
		} ;

	if (outfilename [0] == '-')
	{	printf ("Error : Output filename (%s) looks like an option.\n\n", outfilename) ;
		usage_exit (progname) ;
		} ;

	for (k = 1 ; k < argc - 2 ; k++)
	{	if (! strcmp (argv [k], "-pcms8"))
		{	outfileminor = SF_FORMAT_PCM_S8 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-pcmu8"))
		{	outfileminor = SF_FORMAT_PCM_U8 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-pcm16"))
		{	outfileminor = SF_FORMAT_PCM_16 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-pcm24"))
		{	outfileminor = SF_FORMAT_PCM_24 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-pcm32"))
		{	outfileminor = SF_FORMAT_PCM_32 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-float32"))
		{	outfileminor = SF_FORMAT_FLOAT ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-float64"))
		{	outfileminor = SF_FORMAT_DOUBLE ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-ulaw"))
		{	outfileminor = SF_FORMAT_ULAW ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-alaw"))
		{	outfileminor = SF_FORMAT_ALAW ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-alac16"))
		{	outfileminor = SF_FORMAT_ALAC_16 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-alac20"))
		{	outfileminor = SF_FORMAT_ALAC_20 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-alac24"))
		{	outfileminor = SF_FORMAT_ALAC_24 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-alac32"))
		{	outfileminor = SF_FORMAT_ALAC_32 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-ima-adpcm"))
		{	outfileminor = SF_FORMAT_IMA_ADPCM ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-ms-adpcm"))
		{	outfileminor = SF_FORMAT_MS_ADPCM ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-gsm610"))
		{	outfileminor = SF_FORMAT_GSM610 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-dwvw12"))
		{	outfileminor = SF_FORMAT_DWVW_12 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-dwvw16"))
		{	outfileminor = SF_FORMAT_DWVW_16 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-dwvw24"))
		{	outfileminor = SF_FORMAT_DWVW_24 ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-vorbis"))
		{	outfileminor = SF_FORMAT_VORBIS ;
			continue ;
			} ;
		if (! strcmp (argv [k], "-opus"))
		{	outfileminor = SF_FORMAT_OPUS ;
			continue ;
			} ;

		if (strstr (argv [k], "-override-sample-rate=") == argv [k])
		{	const char *ptr ;

			ptr = argv [k] + strlen ("-override-sample-rate=") ;
			override_sample_rate = atoi (ptr) ;
			continue ;
			} ;

		if (! strcmp (argv [k], "-endian=little"))
		{	endian = SF_ENDIAN_LITTLE ;
			continue ;
			} ;

		if (! strcmp (argv [k], "-endian=big"))
		{	endian = SF_ENDIAN_BIG ;
			continue ;
			} ;

		if (! strcmp (argv [k], "-endian=cpu"))
		{	endian = SF_ENDIAN_CPU ;
			continue ;
			} ;

		if (! strcmp (argv [k], "-endian=file"))
		{	endian = SF_ENDIAN_FILE ;
			continue ;
			} ;

		if (! strcmp (argv [k], "-normalize"))
		{	normalize = SF_TRUE ;
			continue ;
			} ;

		printf ("Error : Not able to decode argunment '%s'.\n", argv [k]) ;
		exit (1) ;
		} ;

	memset (&sfinfo, 0, sizeof (sfinfo)) ;

	if ((infile = sf_open (infilename, SFM_READ, &sfinfo)) == NULL)
	{	printf ("Not able to open input file %s.\n", infilename) ;
		puts (sf_strerror (NULL)) ;
		return 1 ;
		} ;

	/* Update sample rate if forced to something else. */
	if (override_sample_rate)
		sfinfo.samplerate = override_sample_rate ;

	infileminor = sfinfo.format & SF_FORMAT_SUBMASK ;

	if ((sfinfo.format = sfe_file_type_of_ext (outfilename, sfinfo.format)) == 0)
	{	printf ("Error : Not able to determine output file type for %s.\n", outfilename) ;
		return 1 ;
		} ;

	outfilemajor = sfinfo.format & (SF_FORMAT_TYPEMASK | SF_FORMAT_ENDMASK) ;

	if (outfileminor == 0)
		outfileminor = sfinfo.format & SF_FORMAT_SUBMASK ;

	if (outfileminor != 0)
		sfinfo.format = outfilemajor | outfileminor ;
	else
		sfinfo.format = outfilemajor | (sfinfo.format & SF_FORMAT_SUBMASK) ;

	sfinfo.format |= endian ;

	if ((sfinfo.format & SF_FORMAT_TYPEMASK) == SF_FORMAT_XI)
		switch (sfinfo.format & SF_FORMAT_SUBMASK)
		{	case SF_FORMAT_PCM_16 :
					sfinfo.format = outfilemajor | SF_FORMAT_DPCM_16 ;
					break ;

			case SF_FORMAT_PCM_S8 :
			case SF_FORMAT_PCM_U8 :
					sfinfo.format = outfilemajor | SF_FORMAT_DPCM_8 ;
					break ;
			} ;

	if (sf_format_check (&sfinfo) == 0)
	{	sf_close (infile) ;
		report_format_error_exit (argv [0], &sfinfo) ;
		} ;

	if ((sfinfo.format & SF_FORMAT_SUBMASK) == SF_FORMAT_GSM610 && sfinfo.samplerate != 8000)
	{	printf (
			"WARNING: GSM 6.10 data format only supports 8kHz sample rate. The converted\n"
			"ouput file will contain the input data converted to the GSM 6.10 data format\n"
			"but not re-sampled.\n"
			) ;
		} ;

	/* Open the output file. */
	if ((outfile = sf_open (outfilename, SFM_WRITE, &sfinfo)) == NULL)
	{	printf ("Not able to open output file %s : %s\n", outfilename, sf_strerror (NULL)) ;
		return 1 ;
		} ;

	/* Copy the metadata */
	copy_metadata (outfile, infile, sfinfo.channels) ;

	if (normalize
			|| (outfileminor == SF_FORMAT_DOUBLE) || (outfileminor == SF_FORMAT_FLOAT)
			|| (infileminor == SF_FORMAT_DOUBLE) || (infileminor == SF_FORMAT_FLOAT)
			|| (infileminor == SF_FORMAT_OPUS) || (outfileminor == SF_FORMAT_OPUS)
			|| (infileminor == SF_FORMAT_VORBIS) || (outfileminor == SF_FORMAT_VORBIS))
		sfe_copy_data_fp (outfile, infile, sfinfo.channels, normalize) ;
	else
		sfe_copy_data_int (outfile, infile, sfinfo.channels) ;

	sf_close (infile) ;
	sf_close (outfile) ;

	return 0 ;
} /* main */

static void
copy_metadata (SNDFILE *outfile, SNDFILE *infile, int channels)
{	SF_INSTRUMENT inst ;
	SF_CUES cues ;
	SF_BROADCAST_INFO_2K binfo ;
	const char *str ;
	int k, chanmap [256] ;

	for (k = SF_STR_FIRST ; k <= SF_STR_LAST ; k++)
	{	str = sf_get_string (infile, k) ;
		if (str != NULL)
			sf_set_string (outfile, k, str) ;
		} ;

	memset (&inst, 0, sizeof (inst)) ;
	memset (&cues, 0, sizeof (cues)) ;
	memset (&binfo, 0, sizeof (binfo)) ;

	if (channels < ARRAY_LEN (chanmap))
	{	size_t size = channels * sizeof (chanmap [0]) ;

		if (sf_command (infile, SFC_GET_CHANNEL_MAP_INFO, chanmap, size) == SF_TRUE)
			sf_command (outfile, SFC_SET_CHANNEL_MAP_INFO, chanmap, size) ;
		} ;

	if (sf_command (infile, SFC_GET_CUE, &cues, sizeof (cues)) == SF_TRUE)
		sf_command (outfile, SFC_SET_CUE, &cues, sizeof (cues)) ;

	if (sf_command (infile, SFC_GET_INSTRUMENT, &inst, sizeof (inst)) == SF_TRUE)
		sf_command (outfile, SFC_SET_INSTRUMENT, &inst, sizeof (inst)) ;

	if (sf_command (infile, SFC_GET_BROADCAST_INFO, &binfo, sizeof (binfo)) == SF_TRUE)
		sf_command (outfile, SFC_SET_BROADCAST_INFO, &binfo, sizeof (binfo)) ;

} /* copy_metadata */

